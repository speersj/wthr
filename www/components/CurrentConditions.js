import React from 'react'
import PropTypes from 'prop-types'
import { Heading, Text } from 'rebass'
import WeatherIcon from './WeatherIcon'
import { isNotEmpty } from '../lib/utils'
import Temperature from './Temperature'
import TextBoxCentered from './TextBoxCentered'

CurrentConditions.propTypes = {
  forecastSummary: PropTypes.string,
  conditions: PropTypes.shape({
    summary: PropTypes.string,
    icon: PropTypes.string,
    temperature: PropTypes.number,
    apparentTemperature: PropTypes.number,
  }),
}

export default function CurrentConditions(props) {
  const isLoaded = isNotEmpty(props.conditions)
  const { conditions } = props
  return (
    <TextBoxCentered py={4} width={1} mx="auto">
      <Heading py={3} fontSize={[5, 6]} color={conditions && conditions.icon}>
        {isLoaded ? <WeatherIcon name={conditions.icon} /> : 'ðŸŒˆ'}
        <span data-testid="test-temp">
          {' '}
          {isLoaded && (
            <Temperature
              value={conditions.temperature}
              color={conditions.icon || 'textEm'}
            />
          )}
        </span>
      </Heading>
      <Text py={3} fontSize={[2, 3]} color="accent2">
        {isLoaded && `${conditions.summary}, feels like `}{' '}
        {isLoaded && (
          <Temperature value={conditions.apparentTemperature} inline />
        )}
      </Text>
      <Text fontSize={[3, 4]} color="accent">
        {props.forecastSummary}
      </Text>
    </TextBoxCentered>
  )
}
