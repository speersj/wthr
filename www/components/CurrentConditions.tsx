import React from "react";
import { Heading, Text } from "rebass";
import Temperature from "./Temperature";
import TextBoxCentered from "./TextBoxCentered";
import WeatherIcon from "./WeatherIcon";
import { ForecastAPICurrently } from "../lib/ForecastAPIResponse";

interface Props {
  forecastSummary: string;
  conditions: ForecastAPICurrently;
}

const CurrentConditions = (props: Props) => {
  const isLoaded = props.forecastSummary && props.forecastSummary.length > 0;
  const { conditions } = props;
  return (
    <TextBoxCentered px={3} py={4} width={1} mx="auto" bg="bg">
      <Heading py={3} fontSize={[5, 6]} color={conditions && conditions.icon}>
        {isLoaded ? <WeatherIcon name={conditions.icon} /> : "ðŸŒˆ"}
        <span data-testid="test-temp">
          {" "}
          {isLoaded && (
            <Temperature
              value={conditions.temperature}
              color={conditions.icon || "textEm"}
            />
          )}
        </span>
      </Heading>
      <Text py={3} fontSize={[2, 3]} color="accent2">
        {isLoaded && `${conditions.summary}, feels like `}{" "}
        {isLoaded && (
          <Temperature value={conditions.apparentTemperature} inline />
        )}
      </Text>
      <Text fontSize={[3, 4]} color="accent">
        {props.forecastSummary}
      </Text>
    </TextBoxCentered>
  );
};

export default CurrentConditions;
