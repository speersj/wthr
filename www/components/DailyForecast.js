import React from 'react'
import PropTypes from 'prop-types'
import { Flex, Box, Text } from 'rebass'
import WeatherIcon from './WeatherIcon'
import { toDayOfWeek, toDate, toTime } from '../lib/dateTime'
import Temperature from './Temperature'
import TextCentered from './TextCentered'
import Title from './Title'
import TitleBig from './TitleBig'
import TextBoxCentered from './TextBoxCentered'
import Icon from './Icon'
import MoonPhaseIcon, { moonPhaseName } from './MoonPhase'
import TextSmall from './TextSmall'
import RotatedSpan from './RotatedSpan'
import Percentage from './Percentage'
import UVIndex from './UVIndex'
import { capitalize } from '../lib/utils'

DailyForecast.propTypes = {
  time: PropTypes.number, // unix time
  summary: PropTypes.string,
  icon: PropTypes.string,

  temperatureLow: PropTypes.number,
  temperatureHigh: PropTypes.number,

  sunriseTime: PropTypes.number, // unix time
  sunsetTime: PropTypes.number, // unix time
  moonPhase: PropTypes.number,

  precipProbability: PropTypes.number, // percentage
  precipIntensity: PropTypes.number, // inches / hour
  precipType: PropTypes.string, // rain / snow etc

  windSpeed: PropTypes.number, // mph
  windBearing: PropTypes.number, // direction 0->360
  cloudCover: PropTypes.number, // percentage

  dewPoint: PropTypes.number, // deg f
  humidity: PropTypes.number, // percentage
  pressure: PropTypes.number, // millibars

  uvIndex: PropTypes.number,
  visibility: PropTypes.number, // up to 10 miles
}

const Cell = props => (
  <TextBoxCentered width={1 / 3} fontSize={[3, 4]} {...props} my={3}>
    {props.children}
  </TextBoxCentered>
)

Cell.propTypes = { children: PropTypes.node }

export default function DailyForecast(props) {
  const { time } = props

  if (time == undefined) {
    return (
      <Box width={1} bg="bgEm">
        <TitleBig>üåà</TitleBig>
      </Box>
    )
  }

  return (
    <Box width={1} bg="bgEm">
      <Title color="text">{`${toDayOfWeek(time)}, ${toDate(time)}`}</Title>
      <TitleBig>
        <WeatherIcon name={props.icon} />
      </TitleBig>

      <TextBoxCentered p={2} mb={4} fontSize={[2, 3]}>
        <Temperature value={props.temperatureHigh} color="hot" inline />{' '}
        <Temperature value={props.temperatureLow} color="cold" inline />
        <TextCentered fontSize={[2, 3]} my={2} color="accent2">
          {props.summary}
        </TextCentered>
        {props.precipProbability === 0 ? (
          <Text mt={2}>No Precipitation</Text>
        ) : (
          <>
            <Text mt={2} color="rainy">
              <Percentage value={props.precipProbability} inline /> Chance of{' '}
              {capitalize(props.precipType || 'Precipitation')},{' '}
              {(props.precipIntensity * 24).toFixed(2)} in
            </Text>
          </>
        )}
        <Text color="cloudy" mt={2}>
          <Percentage value={props.cloudCover} inline /> Cloudy
        </Text>
      </TextBoxCentered>

      <Flex flexWrap="wrap" justifyContent="space-evenly" alignItems="center">
        <Cell width={1 / 3} color="orange">
          <Icon name="sunrise" />
          <TextSmall>{toTime(props.sunriseTime)}</TextSmall>
        </Cell>

        <Cell width={1 / 3} color="textMuted">
          <Icon name="sunset" />
          <TextSmall>{toTime(props.sunsetTime)}</TextSmall>
        </Cell>

        <Cell width={1 / 3} color="contrast">
          <span data-testid="moon-phase-icon">
            <MoonPhaseIcon moonPhase={props.moonPhase} />
          </span>
          <TextSmall>{moonPhaseName(props.moonPhase)}</TextSmall>
        </Cell>

        <Cell width={1 / 3}>
          <TextSmall>Wind</TextSmall>
          <TextSmall>
            {Math.round(props.windSpeed)}mph
            <RotatedSpan angle={props.windBearing || 0}> ‚¨Ü Ô∏è</RotatedSpan>
          </TextSmall>
        </Cell>

        <Cell width={1 / 3}>
          <TextSmall>
            Dew Point
            <Temperature value={props.dewPoint} />
          </TextSmall>
        </Cell>

        <Cell width={1 / 3}>
          <TextSmall>Humidity</TextSmall>
          <TextSmall>
            <Percentage value={props.humidity} />
          </TextSmall>
        </Cell>

        <Cell width={1 / 3}>
          <TextSmall>Pressure</TextSmall>
          <TextSmall>{Math.round(props.pressure)}mb</TextSmall>
        </Cell>

        <Cell width={1 / 3}>
          <TextSmall>Visibility</TextSmall>
          <TextSmall>{visibilityText(props.visibility)}</TextSmall>
        </Cell>

        <Cell width={1 / 3}>
          <TextSmall>
            <UVIndex uvIndex={props.uvIndex} />
          </TextSmall>
        </Cell>
      </Flex>
    </Box>
  )
}

function visibilityText(visibility) {
  const vis = Math.round(visibility)
  return vis >= 10 ? '> 10 miles' : `${vis} miles`
}
