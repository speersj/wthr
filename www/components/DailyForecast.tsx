import React from "react";
import { Box, Flex, Text } from "rebass";
import { toDate, toDayOfWeek, toTime } from "../lib/dateTime";
import { capitalize } from "../lib/utils";
import Cell from "./DailyForecastCell";
import Icon from "./Icon";
import MoonPhaseIcon, { moonPhaseName } from "./MoonPhase";
import Percentage from "./Percentage";
import RotatedSpan from "./RotatedSpan";
import Temperature from "./Temperature";
import TextBoxCentered from "./TextBoxCentered";
import TextCentered from "./TextCentered";
import TextSmall from "./TextSmall";
import Title from "./Title";
import TitleBig from "./TitleBig";
import UVIndex from "./UVIndex";
import WeatherIcon from "./WeatherIcon";

export interface Props {
  time: number; // unix time
  summary: string;
  icon: string;

  temperatureLow: number;
  temperatureHigh: number;

  sunriseTime: number; // unix time
  sunsetTime: number; // unix time
  moonPhase: number;

  precipProbability: number; // percentage
  precipIntensity: number; // inches / hour
  precipType: string; // rain / snow etc

  windSpeed: number; // mph
  windBearing: number; // direction 0->360
  cloudCover: number; // percentage

  dewPoint: number; // deg f
  humidity: number; // percentage
  pressure: number; // millibars

  uvIndex: number;
  visibility: number; // up to 10 miles
}

const DailyForecast = (props: Props) => {
  const { time } = props;

  if (!time && time !== 0) {
    return (
      <Box width={1} bg="bgEm">
        <TitleBig>üåà</TitleBig>
      </Box>
    );
  }

  return (
    <Box width={1} bg="bgEm">
      <Title color="text">{`${toDayOfWeek(time)}, ${toDate(time)}`}</Title>
      <TitleBig>
        <WeatherIcon name={props.icon} />
      </TitleBig>

      <TextBoxCentered p={2} mb={4} fontSize={[2, 3]}>
        <Temperature value={props.temperatureHigh} color="hot" inline />{" "}
        <Temperature value={props.temperatureLow} color="cold" inline />
        <TextCentered fontSize={[2, 3]} my={2} color="accent2">
          {props.summary}
        </TextCentered>
        {props.precipProbability === 0 ? (
          <Text mt={2}>No Precipitation</Text>
        ) : (
          <>
            <Text mt={2} color="rainy">
              <Percentage value={props.precipProbability} inline /> Chance of{" "}
              {capitalize(props.precipType || "Precipitation")},{" "}
              {(props.precipIntensity * 24).toFixed(2)} in
            </Text>
          </>
        )}
        <Text color="cloudy" mt={2}>
          <Percentage value={props.cloudCover} inline /> Cloudy
        </Text>
      </TextBoxCentered>

      <Flex flexWrap="wrap" justifyContent="space-evenly" alignItems="center">
        <Cell width={1 / 3} color="orange">
          <Icon name="sunrise" />
          <TextSmall>{toTime(props.sunriseTime)}</TextSmall>
        </Cell>

        <Cell width={1 / 3} color="textMuted">
          <Icon name="sunset" />
          <TextSmall>{toTime(props.sunsetTime)}</TextSmall>
        </Cell>

        <Cell width={1 / 3} color="contrast">
          <span data-testid="moon-phase-icon">
            <MoonPhaseIcon moonPhase={props.moonPhase} />
          </span>
          <TextSmall>{moonPhaseName(props.moonPhase)}</TextSmall>
        </Cell>

        <Cell width={1 / 3}>
          <TextSmall>Wind</TextSmall>
          <TextSmall>
            {Math.round(props.windSpeed)}mph
            <RotatedSpan angle={props.windBearing || 0}> ‚¨Ü Ô∏è</RotatedSpan>
          </TextSmall>
        </Cell>

        <Cell width={1 / 3}>
          <TextSmall>
            Dew Point
            <Temperature value={props.dewPoint} />
          </TextSmall>
        </Cell>

        <Cell width={1 / 3}>
          <TextSmall>Humidity</TextSmall>
          <TextSmall>
            <Percentage value={props.humidity} />
          </TextSmall>
        </Cell>

        <Cell width={1 / 3}>
          <TextSmall>Pressure</TextSmall>
          <TextSmall>{Math.round(props.pressure)}mb</TextSmall>
        </Cell>

        <Cell width={1 / 3}>
          <TextSmall>Visibility</TextSmall>

          <TextSmall>{renderVisibility(props.visibility)}</TextSmall>
        </Cell>

        <Cell width={1 / 3}>
          <TextSmall>
            <UVIndex uvIndex={props.uvIndex} />
          </TextSmall>
        </Cell>
      </Flex>
    </Box>
  );
};

function renderVisibility(visibility: number) {
  const vis = Math.round(visibility);
  return vis >= 10 ? "> 10 miles" : `${vis} miles`;
}

export default DailyForecast;
