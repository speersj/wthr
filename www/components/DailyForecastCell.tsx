import React from "react";
import TextBoxCentered from "./TextBoxCentered";
import { BoxProps } from "rebass";

export default function DailyForecastCell(props: BoxProps) {
  return (
    <TextBoxCentered width={1 / 3} fontSize={[3, 4]} my={3} {...props}>
      {props.children}
    </TextBoxCentered>
  );
}
