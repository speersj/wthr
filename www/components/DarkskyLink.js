import React from 'react'
import styled from 'styled-components'
import { Link } from 'rebass'

const StyledLink = styled(Link)`
  text-decoration: none;

  :hover {
    color: black;
    background-color: yellow;
  }
`

export default function DarkSkyLink(props) {
  return (
    <StyledLink
      {...props}
      href="https://darksky.net/poweredby/"
      m={1}
      p={2}
      target="_blank"
      rel="noopener noreferrer"
    >
      Powered by Dark Sky
    </StyledLink>
  )
}
