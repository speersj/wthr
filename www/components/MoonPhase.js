import React from 'react'
import PropTypes from 'prop-types'
import { capitalize } from '../lib/utils'

MoonPhaseIcon.propTypes = { moonPhase: PropTypes.number }
MoonPhaseIcon.defaultProps = { moonPhase: 0 }

/**
 *
 * @param {Number} moonPhase - 0 -> 1 moonPhase from darksky api
 */
export default function MoonPhaseIcon({ moonPhase }) {
  const title = moonPhaseName(moonPhase) + ' Moon'
  return (
    <i title={title} role="img" className={moonPhaseClassName(moonPhase)} />
  )
}

/**
 *
 * @param {number} moonPhase value from darkSky API
 * @returns {string} name of moon phase, ex 'Waxing Crescent', 'Full', etc
 */
export function moonPhaseName(moonPhase) {
  return moonPhaseClassName(moonPhase)
    .replace('wi wi-moon-', '')
    .replace(/-\d/, '')
    .split('-')
    .map(capitalize)
    .join(' ')
}

/**
 *
 * @param {number} moonPhase
 * @returns {string} class name for displaying weather icon
 */
export function moonPhaseClassName(moonPhase) {
  return (
    'wi ' +
    [
      'wi-moon-new',
      'wi-moon-waxing-crescent-1',
      'wi-moon-waxing-crescent-2',
      'wi-moon-waxing-crescent-3',
      'wi-moon-waxing-crescent-4',
      'wi-moon-waxing-crescent-5',
      'wi-moon-waxing-crescent-6',

      'wi-moon-first-quarter',
      'wi-moon-waxing-gibbous-1',
      'wi-moon-waxing-gibbous-2',
      'wi-moon-waxing-gibbous-3',
      'wi-moon-waxing-gibbous-4',
      'wi-moon-waxing-gibbous-5',
      'wi-moon-waxing-gibbous-6',

      'wi-moon-full',
      'wi-moon-waning-gibbous-1',
      'wi-moon-waning-gibbous-2',
      'wi-moon-waning-gibbous-3',
      'wi-moon-waning-gibbous-4',
      'wi-moon-waning-gibbous-5',
      'wi-moon-waning-gibbous-6',

      'wi-moon-third-quarter',
      'wi-moon-waning-crescent-1',
      'wi-moon-waning-crescent-2',
      'wi-moon-waning-crescent-3',
      'wi-moon-waning-crescent-4',
      'wi-moon-waning-crescent-5',
      'wi-moon-waning-crescent-6',
    ][Math.floor(moonPhase * 28)]
  )
}
