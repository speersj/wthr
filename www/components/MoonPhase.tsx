import React from "react";

import { capitalize } from "../lib/utils";

/**
 * returns name of icon for given moon phase (0 to 1)
 */
export default function MoonPhaseIcon({ moonPhase }: { moonPhase: number }) {
  const title = moonPhaseName(moonPhase) + " Moon";
  return (
    <i title={title} role="img" className={moonPhaseClassName(moonPhase)} />
  );
}

/**
 * given moonPhase value from darkSky API, returns name of that
 * phase "Waxing Crescent", "Full", etc.
 */
export function moonPhaseName(moonPhase: number) {
  return moonPhaseClassName(moonPhase)
    .replace("wi wi-moon-", "")
    .replace(/-\d/, "")
    .split("-")
    .map(capitalize)
    .join(" ");
}

/**
 * returns class name for i element given a moonPhase from darksky api
 */
export function moonPhaseClassName(moonPhase: number) {
  return (
    "wi " +
    [
      "wi-moon-new",
      "wi-moon-waxing-crescent-1",
      "wi-moon-waxing-crescent-2",
      "wi-moon-waxing-crescent-3",
      "wi-moon-waxing-crescent-4",
      "wi-moon-waxing-crescent-5",
      "wi-moon-waxing-crescent-6",

      "wi-moon-first-quarter",
      "wi-moon-waxing-gibbous-1",
      "wi-moon-waxing-gibbous-2",
      "wi-moon-waxing-gibbous-3",
      "wi-moon-waxing-gibbous-4",
      "wi-moon-waxing-gibbous-5",
      "wi-moon-waxing-gibbous-6",

      "wi-moon-full",
      "wi-moon-waning-gibbous-1",
      "wi-moon-waning-gibbous-2",
      "wi-moon-waning-gibbous-3",
      "wi-moon-waning-gibbous-4",
      "wi-moon-waning-gibbous-5",
      "wi-moon-waning-gibbous-6",

      "wi-moon-third-quarter",
      "wi-moon-waning-crescent-1",
      "wi-moon-waning-crescent-2",
      "wi-moon-waning-crescent-3",
      "wi-moon-waning-crescent-4",
      "wi-moon-waning-crescent-5",
      "wi-moon-waning-crescent-6",
    ][Math.floor(moonPhase * 28)]
  );
}
