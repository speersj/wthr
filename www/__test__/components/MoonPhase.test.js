import { moonPhaseClassName, moonPhaseName } from '../../components/MoonPhase'

describe('moonPhaseClassName', () => {
  it('returns "wi wi-moon-new" for a zero value', () => {
    expect(moonPhaseClassName(0)).toEqual('wi wi-moon-new')
  })

  it('returns "wi wi-moon-full" for a 0.5 value', () => {
    expect(moonPhaseClassName(0.5)).toEqual('wi wi-moon-full')
  })
})

describe('moonPhaseName', () => {
  it('returns new for a new moon (0) value', () => {
    expect(moonPhaseName(0)).toEqual('New')
  })

  it('returns waxing crescent for < 0.25', () => {
    expect(moonPhaseName(0.15)).toEqual('Waxing Crescent')
  })
})
