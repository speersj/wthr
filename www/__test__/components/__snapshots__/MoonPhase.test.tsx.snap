// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoonPhaseIcon renders correct html 1`] = `
<i
  class="wi wi-moon-waxing-crescent-5"
  role="img"
  title="Waxing Crescent Moon"
/>
`;
